lang/LangPrimSource/SC_ComPort.cpp

```diff
@@ -111,127 +251,91 @@ void SC_UdpInPort::handleReceivedUDP(const boost::system::error_code& error, std
     }
 
     if (error) {
-        printf("SC_UdpInPort: received error - %s\n", error.message().c_str());
+        printf("UDP: received error - %s\n", error.message().c_str());
         startReceiveUDP();
         return;
     }
 
-    OSC_Packet* packet = (OSC_Packet*)malloc(sizeof(OSC_Packet));
-
-    packet->mReplyAddr.mProtocol = kUDP;
-    packet->mReplyAddr.mAddress = remoteEndpoint.address();
-    packet->mReplyAddr.mPort = remoteEndpoint.port();
-    packet->mReplyAddr.mSocket = udpSocket.native_handle();
+    std::unique_ptr<char[]> data(new char[bytesTransferred]);
+    memcpy(data.get(), mRecvBuffer.data(), bytesTransferred);
 
-    char* data = (char*)malloc(bytesTransferred);
-    packet->mSize = bytesTransferred;
-    packet->mData = data;
-    memcpy(data, recvBuffer.data(), bytesTransferred);
+    mHandleFunc(std::move(data), bytesTransferred);
 
-    ProcessOSCPacket(packet, mPortNum, timeReceived);
     startReceiveUDP();
 }
 
 //////////////////////////////////////////////////////////////////////////////////////////////////////////
 
-SC_UdpCustomInPort::SC_UdpCustomInPort(int inPortNum): SC_UdpInPort(inPortNum, 1) {}
-
-SC_UdpCustomInPort::~SC_UdpCustomInPort() {}
+UDPCustom::UDPCustom(int inPortNum, HandlerType handlerType): UDP(inPortNum, handlerType, 1) {}
 
 //////////////////////////////////////////////////////////////////////////////////////////////////////////
 
-SC_TcpInPort::SC_TcpInPort(int inPortNum, int inMaxConnections, int inBacklog):
-    acceptor(ioService, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), inPortNum)),
-    mPortNum(inPortNum) {
+TCP::TCP(int inPortNum, int inMaxConnections, int inBacklog, HandlerType handlerType):
+    mAcceptor(ioService, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), inPortNum)),
+    mPortNum(inPortNum),
+    mHandlerType(handlerType) {
     // FIXME: handle max connections
     // FIXME: backlog???
 
     startAccept();
 }
 
-void SC_TcpInPort::startAccept() {
-    SC_TcpConnection::pointer newConnection(new SC_TcpConnection(ioService, this));
+void TCP::startAccept() {
+    Detail::TCPConnection::pointer newConnection(new Detail::TCPConnection(ioService, mPortNum, mHandlerType));
```

Comment:
I think `std::make_shared` makes this more clear - and given that we aren't modifying the `shared_ptr` itself, from what I can see it can maybe even be const?

```suggestion
    const auto newConnection = std::make_shared<Detail::TCPConnection>(ioService, mPortNum, mHandlerType);
```